### PR TITLE
[prototype only] Add cutlass as an alternative backend to inductor

### PR DESCRIPTION
**Pull request by ipiszy**
**NOTE: This PR is a prototype to collect early feedback. I'll follow up on splitting it into small PRs and add proper unittests.**

This PR adds cutlass support for simple gemm cases (matmul and linear), without arbitrary epilogue fusions. More support for epilogue fusion will be implemented on top of cutlass epilogue visitor, which is to be released in cutlass 3.2.

### Key changes:
* API:
Introduce "Cutlass" as an alternative backend in `max_autotune_gemm_backends` in torch/_inductor/config.py.
This option is exposed via `torch.compile()` API `options` arg.

* CUDA env:
Add a file torch/_inductor/codegen/cuda/cuda_env.py to get cuda version / arch. Not sure whether there are better ways to do this in Pytorch.

* CUDA / CutlassGemm template:
  * IR: Create `TritonTemplateBuffer` and `CUDATemplateBuffer` subclasses. CUDATemplateBuffer keeps workspace_size which is required by cutlass kernels. Remembering the workspace_size is useful for memory planning.
  * Templates: Create `CUDATemplate` and `CutlassGemmTemplate` for code rendering.
  * Kernel: Create `CUDAKernel` and `CUDATemplateKernel` for code rendering.
  * Scheduling: Create `CUDAScheduling`, which simply inherit from `TritonScheduling`, since it's only used by template codegen instead of more general node fusions. Accordingly, rely on "triton.py" for `codegen_template()` and `can_fuse()` and add simple `if / else` branches for CUDAScheduling. This looks a bit hacky, please suggest better ways.
  * Wrapper: Add changes to support calling from Python to methods defined in .so. Also update memory_planning part to take "workspace_size" into consideration.
 
* CUDA codecache / autotuning
  * Compilation: Add `CUDACodeCache` for CUDA compilation. Currently it relies on nvcc. More follow-ups are needed, e.g. nvrtc, add support for non-linux platform, and add support for internal production environment.
  * Max autotune: Add `TritonBenchmarkRequest` and `CUDABenchmarkRequest` subclasses to handle benchmarking.

* Profiling util
  * I found that current profiling time is very inaccurate as CPU-side overhead is dominant. This PR tries to rely on `torch.profile()` to get accurate timing info. However, one caveat is that it may exclude `vectorized_elementwise_kernel` kernel incorrectly which is used to clear L2 cache.

### Tests:
* Gemm bias / Gemm + matrix tests * A100 / H100 tests all done.
Static shape only. 

### Things to fix in this PR:
* Update max autotune cache file format to also record "workspace_size" for each CUDA kernel. Otherwise if workspace_size > 0 and cache hits, current logic will break.

### Things to fix in follow-up PRs:
* Support AOTInductor.
* Support dynamic shapes.
* Make cutlass profiling more robust. It's possible that selected cutlass configs don't compile. We should either select configs more smartly, or avoid crashing the current process when compilation fails.
* Dedup cutlass configs. It's way too many now.
* Parallel compilation to improve compilation speed. Now it's super slow.
* Parallel benchmarking using multi-GPUs.

### Things to add:
* fp8
* epilogue visitor, general epilogue fusion
* conv kernels

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -557,23 +557,23 @@ def cpp_argdefs(self):
                 continue
             outer = inplaced.other_names[-1]
             inner = inplaced.inner_name
-            dtype = buffer_types[outer]
+            dtype = buffer_types.get(outer, V.graph.get_dtype(outer))
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:560]: We might be able to just call `V.graph.get_dtype(outer)` directly and delete `buffer_types` from this function.  I think `get_dtype` already handles constants (or we could easily make it handle them).

File: torch/_inductor/autotune_process.py [hunk 2]
```diff
@@ -266,4 +382,5 @@ def benchmark_in_sub_process(
             # return INF so this choice will be ignored
             return float("inf")
 
+        print(f"benchmark successful! {timing=}")
```
Comment (jansel) [inline torch/_inductor/autotune_process.py:385]: Make sure the final version of PR doesn't print.

File: torch/_inductor/codegen/cuda/cuda_env.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import functools
+from subprocess import PIPE, Popen
+
+from ... import config
+
+
+def _detect_cuda_arch_with_nvidia_smi() -> str:
+    try:
+        proc = Popen(
+            ["nvidia-smi", "--query-gpu=gpu_name", "--format=csv"],
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_env.py:10]: Perhaps:
```
torch.cuda.get_device_properties(0).name
```

File: torch/_inductor/codegen/cuda/cuda_env.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            "75": ["T4", "Quadro T2000"],
+            "80": ["PG509", "A100", "A10G", "RTX 30", "A30", "RTX 40"],
+            "90": ["H100"],
+        }
+        for sm, names in sm_names.items():
+            if any(name in stdout for name in names):
+                return sm
+        return None
+    except Exception:
+        return None
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_env.py:27]: Can we do:
```
major, minor = torch.cuda.get_device_capability(0)
return str(major * 10 + minor)
```

File: torch/_inductor/codegen/cuda/cuda_env.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            stdout=PIPE,
+            stderr=PIPE,
+            shell=True,
+        )
+        stdout, stderr = proc.communicate()
+        stdout = stdout.decode("utf-8")
+        version_str = stdout.split(":")[1].strip()
+        return version_str
+    except Exception:
+        return None
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_env.py:43]: Maybe `torch.version.cuda`?

Though do we need the host side version?

File: torch/_inductor/codegen/cuda/cuda_env.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    return res[1:]
+
+
+def _detect_cuda_arch() -> str:
+    try:
+        from cuda import cuda
+
+        _assert_cuda(cuda.cuInit(0))
+        # Get Compute Capability of the first Visible device
+        major, minor = _assert_cuda(cuda.cuDeviceComputeCapability(0))
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_env.py:58]: Same as torch.cuda.get_device_capability(0)?

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+cexpr = CppPrinter().doprint
+
+
+def _normalize_idx(index: int, total_length: int) -> int:
+    return index if index >= 0 else index + total_length
+
+
+class CUDAKernel(Kernel):
+    overrides = OpOverrides
+    pass
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:24]: pass not needed here

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+
+        res = IndentedBuffer(initial_indent=2)
+        res.tabwidth = 1
+        res.splice(
+            """
+            {{
+              if (!{name_str}) {{
+                int64_t {name_str}_size = {size_str};
+                if ({name_str}_size > 0) {{
+                  throw std::runtime_error("input {name_str} is null!");
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:64]: I think this is check >0, which is different from "null"

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        inputs: List[IRNode],
+        outputs: List[IRNode],
+        names_str: str = "",
+        input_reorder: List[int] = None,
+    ) -> str:
+        """
+        Hook called from template code to generate function def and
+        needed args.
+        """
+        print(f"{input_reorder=}")
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:85]: prints should be removed or become logging calls

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+            )
+        )
+        return res.getvalue()
+
+    def def_kernel(
+        self,
+        inputs: List[IRNode],
+        outputs: List[IRNode],
+        names_str: str = "",
+        input_reorder: List[int] = None,
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:79]: Why is this needed?

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+
+        if node is None:
+            return str(default_value)
+
+        index = _normalize_idx(index, len(node.get_size()))
+        if index < 0:
+            return str(default_value)
+
+        stride = node.get_stride()[index]
+        return cexpr(self.rename_indexing(stride))
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:227]: Could we refactor some of these functions to share more code with TritonTemplateKernel?  If I recall this is duplicated.

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+
+    def row_stride(self, node: IRNode, default_value: int = 0) -> str:
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:229]: It is not clear what this does from the name.

File: torch/_inductor/codegen/cuda/cuda_template.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
+import functools
+import itertools
+from copy import copy
+from typing import List
+from unittest.mock import patch
+
+import sympy
+
+from third_party.cutlass.tools.library import scripts as cutlass_lib
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_template.py:9]: Will need to work out how this import works in OSS

File: torch/_inductor/codegen/cuda/cuda_template.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+from ...ir import Buffer, IRNode, Layout
+from ...utils import IndentedBuffer, unique
+from ...virtualized import V
+from ..common import jinja2_env
+
+from . import cutlass_utils
+from .cuda_kernel import CUDATemplateCaller, CUDATemplateKernel
+
+
+class CUDATemplate:
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_template.py:21]: Is there code that could be shared with TritonTemplate here?

File: torch/_inductor/codegen/cuda/cuda_template.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    def generate(self, **kwargs) -> CUDATemplateCaller:
+        kernel_name = f"cuda_{self.name}"
+        with patch.object(
+            V.graph, "get_dtype", self._fake_get_dtype(self.output_node)
+        ), CUDATemplateKernel(
+            kernel_name=kernel_name,
+        ) as kernel:
+            # need to do call render twice to get all the needed args right
+            self.render(kernel=kernel, **kwargs)
+            code = self.render(kernel=kernel, **kwargs)
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_template.py:63]: I actually removed this from Triton template since it confused both you and Horace, see:
https://github.com/pytorch/pytorch/pull/105987

File: torch/_inductor/codegen/cuda/cuda_template.py [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+                  if (error != cutlass::Status::kSuccess) {                                        \\
+                    auto msg = std::string("[") + __FILE__ + "] Got cutlass error: " +             \\
+                        cutlassGetStatusString(error) + " at: " + std::to_string(__LINE__);        \\
+                    std::cerr << msg << std::endl;                                                 \\
+                    throw std::runtime_error(msg);                                                 \\
+                  }                                                                                \\
+                }
+            """
+        )
+        return res
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_template.py:179]: Perhaps we should refactor all of this header boilerplate into a single `*.h` file we just `#include` from our generated code.

We do something similar for [cpp_prefix.h](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp_prefix.h)

https://github.com/pytorch/pytorch/blob/578969ca618aad8da6e201d40a55b2cc9bfdc16e/torch/_inductor/codecache.py#L921-L928

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 17]
```diff
@@ -0,0 +1,468 @@
...
+{{template.header().getvalue()}}
+
+{{template.globals().getvalue()}}
+
+{{instance_definition}}
+
+// When workspace_size is not a nullptr, populates requested workspace_size and returns.
+// Otherwise, compuates the Gemm kernel using the given workspace ptr.
+extern "C" {
+{{kernel.def_kernel(inputs=[X, W, Bias], outputs=[Y], names_str="X, W, Bias, Y", input_reorder=input_reorder)}} {
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/gemm_template.py:30]: Wont "X"/"W"/etc be undefined here?  Why do we need both inputs and names_str?

File: torch/_inductor/codegen/triton.py [hunk 18]
```diff
@@ -1896,7 +1896,7 @@ def dense_size_str(self):
                 sizes.append("1")
         return f"[{', '.join(sizes)}]"
 
-    def call_kernel(self, name: str):
+    def call_kernel(self, name: str, node: IRNode = None):
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:1899]: Let's make node required and update the users.  We can drop `name` since there is `node.get_name()`.

File: torch/_inductor/codegen/triton.py [hunk 19]
```diff
@@ -2383,11 +2387,11 @@ def codegen_template(self, template_node, epilogue_nodes):
                 node.codegen(kernel.split_and_set_ranges(node.get_ranges()))
 
         # finalize must be called after adding epilogue above
-        src_code = partial_code.finalize()
+        src_code = partial_code if isinstance(partial_code, str) else partial_code.finalize()
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:2390]: Shouldn't be needed, see #105987

File: torch/_inductor/codegen/wrapper.py [hunk 20]
```diff
@@ -762,34 +773,62 @@ def __repr__(self):
...
+        if buffer.get_workspace_size() > 0:
+            workspace_size = tuple(buffer.get_workspace_size())
+            workspace_stride = tuple(1)
+            res.append(
+                f"{buffer.get_name()}_workspace = empty_strided("
+                f"{self.codegen_shape_tuple(workspace_size)}, "
+                f"{self.codegen_shape_tuple(workspace_stride)}, "
+                f"device='{device.type}', dtype={torch.uint8})"
+            )
+        return res
```
Comment (jansel) [inline torch/_inductor/codegen/wrapper.py:797]: Will need to make this work with dynamic shapes.

File: torch/_inductor/codegen/wrapper.py [hunk 21]
```diff
@@ -762,34 +773,62 @@ def __repr__(self):
...
+        res.append(
             f"{self.declare}{new.get_name()} = {self.namespace}as_strided({old.get_name()}, "
             f"{self.codegen_shape_tuple(new.get_size())}, "
-            f"{self.codegen_shape_tuple(new.get_stride())}){del_line}  {self.comment} reuse"
+            f"{self.codegen_shape_tuple(new.get_stride())}){del_line0}  {self.comment} reuse"
         )
+        if new.get_workspace_size() > 0:
+            res.append(f"{self.declare}{new.get_name()}_workspace = {old.get_name()}_workspace{del_line1}  {self.comment} reuse")
+        return res
+
```
Comment (jansel) [inline torch/_inductor/codegen/wrapper.py:831]: I suspect it would be simpler to allocate workspaces separately from regular buffers.   Doing in this way means workspaces can only be reused if the main buffer can be reused (which often not the case).

File: torch/_inductor/utils.py [hunk 22]
```diff
@@ -31,6 +32,64 @@
 VarRanges = Dict[sympy.Expr, sympy.Expr]
 
 
+def do_bench_using_profiling(fn: Callable[[], Any], warmup=25, rep=100) -> None:
```
Comment (jansel) [inline torch/_inductor/utils.py:35]: Can unify the way we benchmark triton and cutlass?

File: torch/_inductor/config.py [hunk 23]
```diff
@@ -73,9 +73,10 @@
 max_autotune_gemm = os.environ.get("TORCHINDUCTOR_MAX_AUTOTUNE_GEMM") == "1"
 
 # Specify candidate backends for gemm autotune.
-# Possible choices are combinations of: ATen, Triton.
+# Possible choices are combinations of: ATen, Triton, Cutlass.
```
Comment (aakhundov) [inline torch/_inductor/config.py:76]: Super-nit: `CUTLASS` looks more authentic.

File: torch/_inductor/config.py [hunk 24]
```diff
@@ -386,6 +387,36 @@ class triton:
...
+class cuda:
+    # Whether to enable custom CUDA kernels.
+    # If True, both Triton and custom CUDA kernel codegen are enabled.
+    # If False, only Triton codegen is enabled.
+    enabled = False
+
+    # CUDA arch to use for CUDA template kernel compilation.
+    # Available options: "70", "75", "80", "90"
+    # When arch is None, the Inductor tries to detect CUDA arch by querying
+    # CUDA Python lib or "nvidia-smi" commandline.
```
Comment (aakhundov) [inline torch/_inductor/config.py:399]: Super-nit: `"nvidia-smi" command`.

File: torch/_inductor/config.py [hunk 25]
```diff
@@ -386,6 +387,36 @@ class triton:
...
+    arch = None
+
+    # CUDA version to use for CUDA template kernel compilation.
+    # e.g. "11.4", "12.1", etc.
+    # When version is None, the Inductor tries to detech CUDA version by
+    # querying CUDA Python lib for "nvidia-smi" commandline.
+    version = None
```
Comment (aakhundov) [inline torch/_inductor/config.py:406]: Is it possible to compile with an earlier CUDA version than installed? If not, should we just auto-detect it from the system in all cases?

File: torch/_inductor/ir.py [hunk 26]
```diff
@@ -1167,6 +1170,11 @@ def loader(idx):
     def get_dtype(self):
         return self.data.get_dtype()
 
+
+    def get_layout(self):
+        return self.data.get_layout()
+
+
```
Comment (aakhundov) [inline torch/_inductor/ir.py:1177]: Nit: extra blank lines above and below. Linter will likely complain.

File: torch/_inductor/ir.py [hunk 27]
```diff
@@ -2072,6 +2083,9 @@ def get_size(self):
     def get_stride(self):
         return list(self.layout.stride)
 
+    def get_workspace_size(self):
+        return 0
+
```
Comment (aakhundov) [inline torch/_inductor/ir.py:2088]: Should we move this to `TemplateBuffer`, as that seems to be the lowest common subclass where this is relevant? Or is it necessary for arbitrary `Buffer` instances to be able to return this?

**Review by aakhundov — COMMENTED** [review]
@ipiszy thanks for the e2e prototype! I've looked through some of the code. Will look further and likely add more comments.

File: torch/_inductor/codegen/wrapper.py [hunk 28]
```diff
@@ -1253,19 +1294,19 @@ def make_buffer_allocation(self, buffer):
             ):
                 return f"auto {name} = outputs[{output_idx}];"
             else:
-                self.outputs_need_copy.add(name)
+                [self.outputs_need_copy.add(name)]
```
Comment (aakhundov) [inline torch/_inductor/codegen/wrapper.py:1297]: `[...]` is not required here, but likely required on the L1295 above?

File: torch/_inductor/kernel/mm.py [hunk 29]
```diff
@@ -174,7 +187,8 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
         else []
     )
     if (
-        inp_expanded.get_stride()[0] == 0
+        use_aten_gemm_kernels()
```
Comment (aakhundov) [inline torch/_inductor/kernel/mm.py:190]: Good catch!

**Comment by ipiszy** [comment]
I've finished basic A100 / H100 gemm, gemm+bias, gemm+matrix tests and fixed some issues. Will work on separate PRs to address comments, thanks!

File: torch/_inductor/codecache.py [hunk 30]
```diff
@@ -1157,6 +1159,207 @@ def load(cls, kernel_name, source_code):
...
+            if hasattr(syms, "dlclose"):
+                f_dlclose = syms.dlclose
+
+        if f_dlclose is not None:
+            f_dlclose.argtypes = [ctypes.c_void_p]
+            f_dlclose(self.DLL._handle)
+        else:
+            logging.warning(
+                "dll unloading function was not found, library may not be unloaded properly!"
+            )
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/codecache.py:1317]: Yeah I copied them from cpp_extension and AITemplate code. After a second thought I think I'll just remove "windows" logics since they're not tested. People could work on more proper window support later.

File: torch/_inductor/codegen/cuda/cuda_env.py (hunk 5 above)
Comment (ipiszy) [reply torch/_inductor/codegen/cuda/cuda_env.py]: Yeah it works! I think we just need the CUDA version.

File: test/inductor/test_max_autotune.py [hunk 31]
```diff
@@ -147,9 +149,66 @@ def mm(a, b):
         a = torch.randn(100, 10).cuda()
         b = torch.randn(10, 100).cuda()
 
-        with config.patch({"max_autotune": True, "autotune_in_subproc": True}):
+        with config.patch(
+            {
+                "max_autotune": True,
+                "autotune_in_subproc": True,
+                "max_autotune_gemm_backends": max_autotune_gemm_backends,
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply test/inductor/test_max_autotune.py:157]: Good catch! I'll just remove it.

File: torch/_inductor/codegen/wrapper.py [hunk 32]
```diff
@@ -762,34 +773,62 @@ def __repr__(self):
...
+        res = []
+        res.append(
             f"{buffer.get_name()} = empty_strided("
             f"{self.codegen_shape_tuple(shape)}, "
             f"{self.codegen_shape_tuple(stride)}, "
             f"device='{device.type}', dtype={dtype})"
         )
+        if buffer.get_workspace_size() > 0:
+            workspace_size = tuple(buffer.get_workspace_size())
+            workspace_stride = tuple(1)
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/codegen/wrapper.py:790]: I'll either remove these logics or add more unittests. Default cutlass gemm kernels only have workspze_size == 0 so in E2E test this path is not exercised. From Haicheng only group_gemm and parallel split_k need this.

**Review by ipiszy — COMMENTED** [review]
Thanks @jansel @aakhundov for the review! I replied some comments, working on PRs now.

File: torch/_inductor/codegen/cuda/cuda_env.py (hunk 5 above)
Comment (aakhundov) [reply torch/_inductor/codegen/cuda/cuda_env.py]: CUDA version the `torch` wheel was compiled with is not necessary the same as on the host? And we're going to rely on the host version to compile CUTLASS code?

File: torch/_inductor/utils.py [hunk 33]
```diff
@@ -31,6 +32,64 @@
...
+    for _ in range(5):
+        cache.zero_()
+        fn()
+    end_event.record()
+    torch.cuda.synchronize()
+    estimate_ms = start_event.elapsed_time(end_event) / 5
+
+    # compute number of warmup and repeat
+    n_warmup = max(1, int(warmup / estimate_ms))
+    n_repeat = max(1, int(rep / estimate_ms))
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/utils.py:53]: I feel it doesn't matter? Is it an issue?

File: torch/_inductor/ir.py (hunk 27 above)
Comment (ipiszy) [reply torch/_inductor/ir.py:2088]: I feel it depends on the algorithm. In theory triton codegen code can also require extra workspace.

File: torch/_inductor/select_algorithm.py [hunk 34]
```diff
@@ -744,15 +732,16 @@ def __init__(self, msg, choice: ChoiceCaller, inputs_str):
...
         if len(choices) == 0:
             raise RuntimeError(
                 "No choices to select, please consider adding ATEN into max_autotune_gemm_backends "
                 "config (defined in torch/_inductor/config.py) to allow at least one choice. "
             )
 
-        if len(choices) == 1:
-            return choices[0].output_node()
+        # if len(choices) == 1:
+        #     return choices[0].output_node()
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/select_algorithm.py:745]: Need to make sure that workspace_size is fetched via the benchmark process.

File: torch/_inductor/codegen/triton.py (hunk 18 above)
Comment (ipiszy) [reply torch/_inductor/codegen/triton.py:1899]: Actually name != node.get_name(). name here is generated by `self.define_kernel(src_code, node_schedule)`.

File: torch/_inductor/utils.py (hunk 33 above)
[the remark this answers is not in the document]
Comment (aakhundov) [reply torch/_inductor/utils.py:53]: Purely from e2e time perspective. By decreasing warmup to, say, 10 iterations, you'll shave off ~1/5 of the total kernel runtime during profiling. Although maybe the total kernel runtime is not a bottleneck, so it doesn't matter indeed.

File: torch/_inductor/ir.py (hunk 27 above)
Comment (aakhundov) [reply torch/_inductor/ir.py:2088]: Also, I think, you're invoking this method on abstract `Buffer` in some context, I remember. So ignore the comment.

File: torch/_inductor/select_algorithm.py (hunk 34 above)
[the remark this answers is not in the document]
Comment (aakhundov) [reply torch/_inductor/select_algorithm.py:745]: I see. I guess, this means that even one-option autotuning (like in the ATen-only cases) will run the kernel / invoke the subprocess etc.? I'm wondering if this can degrade compilation time. Likely the cache will be hit on subsequent attempts, though.

File: torch/_inductor/codecache.py [hunk 35]
```diff
@@ -1303,6 +1533,12 @@ def task():
 
         return self.submit(task)
 
+    def cuda(self, dst_code_suffix, source_code):
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/codecache.py:1537]: I'd like to keep the order here so that it's easier to review generated code. Otherwise the source code part is too large and will make the other params non-visible.

File: torch/_inductor/autotune_process.py [hunk 36]
```diff
@@ -166,67 +168,181 @@ class BenchmarkRequest:
...
+        output_tensor_meta: TensorMeta,
+        extra_args: Dict[str, Any],
+        source_code: str,
+    ):
+        super().__init__(kernel_name, input_tensor_meta, output_tensor_meta, extra_args)
+        self.source_code = source_code
+        self.workspace_size: Optional[int] = None
+        self.DLL: Optional[DLLWrapper] = None
+        self.hash_key: str = ""
+        self.source_file: str = ""
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/autotune_process.py:284]: I'd like to keep type annotation here. The next line doesn't allow me giving type annotations.

**Comment by ipiszy** [comment]
Hi @jansel @aakhundov , I created new PR stack https://github.com/pytorch/pytorch/pull/107933, ptal, thanks!

**Comment by vadimkantorov** [comment]
Related on adding `nvrtc` bindings: 
- https://github.com/pytorch/pytorch/issues/107800
